internal/primitives/state-machine/overlayedchanges/ext.go

```diff
@@ -70,3 +82,713 @@ func compactLen(val uint) int {
+		`target = state 
+		method = ChildGet
+		ext_id = %s
+		child_info = %s
+		key = %s
+		result = %s`,
+		common.BytesToHex(leID(e.Id)),
+		common.BytesToHash(childInfo.StorageKey()),
+		common.BytesToHex(key),
+		common.BytesToHex(result),
+	)
+
+	return result
+}
+
+// Get child storage value hash.
+// This may be optimised for large values.
+// Returns an SCALE encoded hash.
+func (e *Ext[H, Hasher, B]) ChildStorageHash(childInfo storage.ChildInfo, key []byte) []byte {
+	defer guard()
+
+	var hash H
+	result, has := e.overlay.ChildStorage(childInfo, key)
+
+	if has && result == nil {
+		return nil
+	}
+
+	if has && result != nil {
+		hasher := *new(Hasher)
+		hash = hasher.Hash(result)
+	} else {
+		result, err := e.backend.ChildStorageHash(childInfo, key)
+		if err != nil {
+			panic(ExtNotAllowedToFail)
+		}
+
+		if result != nil {
+			hash = *result
+		}
+	}
+
+	logger.Tracef(
+		`target = state 
+		method = ChildHash
+		ext_id = %s
+		child_info = %s
+		key = %s
+		result = %s`,
+		common.BytesToHex(leID(e.Id)),
+		common.BytesToHash(childInfo.StorageKey()),
+		common.BytesToHex(key),
+		hash.String(),
+	)
+
+	if hash.Bytes() != nil {
+		return scale.MustMarshal(hash)
+	}
+
+	return nil
+}
+
+// Whether a storage entry exists.
+func (e *Ext[H, Hasher, B]) ExistsStorage(key []byte) bool {
+	defer guard()
+
+	var exists bool
+
+	value, has := e.overlay.Storage(key)
+	if has {
+		exists = value != nil
+	} else {
+		var err error
+		exists, err = e.backend.ExistsStorage(key)
+		if err != nil {
+			panic(ExtNotAllowedToFail)
+		}
+	}
+
+	logger.Tracef(
+		`target = state 
+		method = Exists
+		ext_id = %s
+		key = %s
+		result = %s`,
+		common.BytesToHex(leID(e.Id)),
+		common.BytesToHex(key),
+		exists,
+	)
+
+	return exists
+}
+
+// Whether a child storage entry exists.
+func (e *Ext[H, Hasher, B]) ExistsChildStorage(childInfo storage.ChildInfo, key []byte) bool {
+	defer guard()
+
+	var exists bool
+
+	value, has := e.overlay.ChildStorage(childInfo, key)
+	if has {
+		exists = value != nil
+	} else {
+		var err error
+		exists, err = e.backend.ExistsChildStorage(childInfo, key)
+		if err != nil {
+			panic(ExtNotAllowedToFail)
+		}
+	}
+
+	logger.Tracef(
+		`target = state 
+		method = ChildExists
+		ext_id = %s
+		child_info = %s
+		key = %s
+		result = %s`,
+		common.BytesToHex(leID(e.Id)),
+		common.BytesToHash(childInfo.StorageKey()),
+		common.BytesToHex(key),
+		exists,
+	)
+
+	return exists
+}
+
+// Returns the key immediately following the given key, if it exists.
+func (e Ext[H, Hasher, B]) NextStorageKey(key []byte) []byte {
+	nextBackendKey, err := e.backend.NextStorageKey(key)
+
+	if err != nil {
+		panic(ExtNotAllowedToFail)
+	}
+
+	overlayedChangesIter := e.overlay.IterAfter(key)
+	overlayChanges := common.NewPeekable2(overlayedChangesIter)
+
+	_, _, has := overlayChanges.Peek()
+
+	if !has {
+		return nextBackendKey
+	}
+
+	if nextBackendKey != nil && has {
+		for overlayKey, overlayValue := range overlayedChangesIter {
```

Comment:
why not range over `overlayChanges` since the `Peeakble2` type supports iteration?